### ISSUE TITLE
Working Rust prototype

### ISSUE THREAD

**Issue by clinuxrulz** [SodiumFRP/sodium#110]
I've been translating a work project (in Java) to Rust recently during my own time at home. And there was some rather complicated user forms that I've used sodium for, and as I translated it to not use FRP I've noticed the final source was hard to maintain (difficult to manage data dependencies).

Out of all the FRP libraries I've used. Sodium was by far the most pleasant. So I tried doing up a Rust implementation.

https://github.com/clinuxrulz/rusty_frp    (lib.rs has a couple of usage examples)

So far I just have:
-  Cells (Behaviours) and no Streams (Events). Streams I plan to handle just as ```type Stream<ENV,A> = Cell<ENV,Option<A>>```
- cell lifts up to artillery 4 (maybe can be unlimited using macros)
- cell map
- cell sink
- cell loop
- transactions

The plan is within a FrpContext, eveything should match Sodium semantics. And as soon as that context's lifetime expires, the whole FRP graph is freed. And Cells can not be mixed between different FrpContexts.

I am missing a cell ```switchC```, I might be able to avoid it using a FrpContext intercommunicating with other FrpContexts. Not sure yet. And the exposed API definitely needs some tweaking.

I'm hoping this prototype will inspire a proper sodium implementation in rust.

**Comment by the-real-blackh** [comment]
This is just what I wanted! You have made me very happy!

Would you like to work on it under the Sodium banner and hopefully get a bit of collaboration going? I've invited you to have commit privilege for the Sodium project. Please create a new project for it, because that's the direction I want to move in - each language being in its own project.

I would like to help but I'm suffering from being overworked at the moment. But I will help at some point in the future!

**Comment by clinuxrulz** [comment]
Its moved over to SodiumFRP/sodium-rust .

Streams (Events) are in, and I believe I've almost got ```switchC``` worked out. ```switchC``` will construct a child graph interconnected with the parent graph and free that child graph when switching to a new child graph (references to cell not allowed to leak outside switch)

What I would be looking for soon are some tests to translate to make sure sodium-rust behaves like sodium is meant to, and keeps that way after refactoring.

**Comment by clinuxrulz** [comment]
cell_switch finished but untested. I saw Denotational.pdf in this repository. I'm thinking of just following it for doing up some tests.

**Comment by the-real-blackh** [comment]
Yes, that would work. Another option would be to use the tests from the Typescript version. I believe they're the tidiest.

**Comment by clinuxrulz** [comment]
Sounds good. Typescript test translation it is then.

**Comment by clinuxrulz** [comment]
Most the test are implemented and working, few more left to translate.

Something is a little off with ```switch_s```: https://github.com/SodiumFRP/sodium-rust/blob/0424d00e6a15cb2fcc3e66668819b8e123e1d857/src/lib.rs#L795

Executing it I get:

```
thread 'tests::switch_s' panicked at 'assertion failed: `(left == right)` (
left: `["A", "B", "C", "d", "e", "F", "G", "h", "I"]`,
right: `["A", "B", "c", "d", "E", "F", "g", "H", "I"]`)', src/lib.rs:826
```

What I've noticed is example:
Let s1 & s2 be Streams, let c be the Cell containing the Stream to switch to (either s1 or s2). And then let s3 be switch_s(c).
If s1 and s2 fire at the same time as c is switching streams, will s3 pass on firing of s1 or s2 at the moment of switching.
It seems the translated typescript test suggest that in the typescript version the previous stream gets fired through s3 rather than the next stream during the switch. However the "Denotational.pdf" suggest the next stream gets fired through s3 during the switch.

In short: The Rust version is following "Denotational.pdf", and the Typescript version is not. The rust version implements ```switch_s``` in terms of ```switch_c``` using the ```type Event<Env,A> = Cell<Env,Option<A>>``` trick. So I have a feeling the Rust version is behaving valid.

However is this merely a pathological case? When your half way between switching between two Streams that are firing at the same time, which one is meant to get passed on? Maybe the it is valid either way as long as it is consistent? (Both Rust and Typescript switch_s are valid)?

Thoughts?

**Comment by the-real-blackh** [comment]
This stuff is really tricky. I'll find some time over the next few days and figure it out.

Thank you so much for your work! I'm really looking forward to playing with it some time and learning Rust properly.

**Comment by clinuxrulz** [comment]
The exposed API will have to be reworked a little bit. If you look at the test code, it feel quite noisy.
I could use some help with restructuring the code to expose an API that has a suitable Fluent Interface feel.

The memory management model I went for is, any new nodes in the graph are kept alive for as long as ```FrpContext``` is alive (with the exception of ```switch``` which will auto-free any nodes created during a switch).
When doing multiple user forms or modes, etc. It is recommended to use a separate ```FrpContext``` for each, then using a parent ```FrpContext``` to stich all the ```FrpContext```s together. This is how to currently avoid a memory leak. (If you used one global ```FrpContext``` for your entire application with dynamic modes or forms, then may have a Space-Time leak, unless your terribly careful)

Not everyone will agree with how I manage the Space-Time, which could lead to two different Rust implementations of Sodium.

With this rust implementation I'm going to aim to avoid ```switch``` myself (```switch``` just provided for a complete Sodium interface). As it seems like a very dynamic graph can be achieved when the nodes of the graph can also be graphs themselves. That is a static graph where the nodes of the graph can change between different static graphs is equal in power to a single dynamic graph.

**Comment by the-real-blackh** [comment]
Just wondering: Does the memory management deal properly with cycles? I remember hearing tell that Rust uses reference counting pointers with a cycle detector.

**Comment by clinuxrulz** [comment]
No idea, I'm pretty new to Rust. I've avoid reference counting to keep the memory management predictable.

**Comment by the-real-blackh** [comment]
OK... Don't understand, but it'll all come out in the wash.

**Comment by clinuxrulz** [comment]
I'll explain a little more without referring to the source code.

The ```FrpContext``` holds the graph which is conceptually ```[(Int,Node)]```, and each ```Node``` contains the dependent nodes by their ID (```Int```).
The ```Node``` is conceptually either a ```Stream``` or ```Cell```.
The end users ```Stream```s and ```Cell```s are just ```Int```s (Node IDs) during runtime. (Which is the same as just handing the end user Weak References, and freeing all data when the ```FrpContext``` gets freed)

Most the graphs I work with are static graphs, so this is OK for me. But I feel like we will just change it to use reference counting so more people will be happy with it.

Because example:
Say you have a Drawing Application that supports having multiple Drawings Open at once.
And lets say the users used a Single FrpContext (instead of one for each drawing tab as recommended)
Each time you start a new drawing tab, more nodes get added to the graph.
When the user closes the tab the nodes are not freed from the graph (because he is using a single FrpContext)
So nodes would keep getting updated even though they will never be observed again. (A Space-Time leak (Which is both a CPU leak and Memory leak at the same time))

For the reason above. We'll need to move over to reference counting ```Rc<...>``` so we do not leak if the users uses a single FrpContext instead of multiple. And using ```Rc<...>``` wont hurt other people using multiple graphs if they wish.

**Comment by clinuxrulz** [comment]
I think I've just realised something. I added the "hold_is_delayed" test, and it was failing. And since "hold" is delayed, and the "switch_s" test uses hold for its stream switching cell. The Typescript version is behaving correctly and my Rust version is not. I completely did not think about "hold" being delayed, sorry about that.

**Comment by the-real-blackh** [comment]
No worries!

Now, about this memory management. We would like memory management to be automatic, because this is one of the powerful things about Sodium. If you want to use using reference counting smart pointers, then to really implement it correctly, we have to deal with cycles. Someone vaguely mentioned to me that Rust can do this. If this is not so, then in the long term we'll need to use the solution that I implemented in the Typescript version. It implements a cycle detector. It was a bit tricky.

C++ doesn't have correct memory management and needs to be fixed. The Typescript implementation would work in C++.

In case you're wondering why Typescript and C++ need to use the same implementation, when one language is garbage collected and the other is not... the reason is that Javascript has no finalizers. This makes the approach used in C# and Java impossible.

**Comment by clinuxrulz** [comment]
I wonder if we can merely make it impossible for the data structures to have a cycle.

```
struct _Cell<Env,A> {
  // points to real cell in graph via ID
  id: u32, 
  // weak reference to graph
  FrpContext<Env,A>
}

impl<Env,A> Drop for _Cell<Env,A> {
  fn drop(self) {
    self.frp_context.report_unreachable_in_userland(self.id);
  }
}

// this one is passed to userland
struct Cell<Env,A> {
  cell: Rc<_Cell<Env,A>>
}
```

This way we prevent cycles in ```Rc``` objects. Then it is up to the ```FrpContext``` to perform a reachability check via the dependences in the nodes (list of node IDs in each node). The typological sort library in use is able to detect cycles, so if we come across a cycle where all nodes are unreachable in userland in the cycles, then they can all be freed together.

**Comment by clinuxrulz** [comment]
Hold on 1 sec. I do not think I have cycles, maybe we can just use ```Rc``` and have no worries.

Cycles will only occur through the use of ```cell_loop``` and ```stream_loop``` right?

Well my ```cell_loop``` uses a fake cycle:

```
    pub fn loop_c<A,F>(&mut self, time0_value: A, k:F) -> Cell<ENV,A>
    where
    A:'static,
    F:Fn(&mut FrpContext<ENV>,&Cell<ENV,A>)->Cell<ENV,A>
    {
        // the fake cell passed to the continuation k.
        let cell = self.new_cell_sink(time0_value);

        // the real cell
        let cell2 = k(self,&Cell::of(cell.id));

        // when ever you try to read the fake cell, you'll get redirected to the real cell.
        self.cell_loop_map.insert(cell.id, cell2.id);

        // return the real cell
        return Cell::of(cell2.id);
    }
```

If you follow the comments in the code above. The internal data structure created within the ```FrpContext``` for the cell_loop has no real cycle, because of the fake cell trick.

**Comment by clinuxrulz** [comment]
Also: https://doc.rust-lang.org/std/rc/

Quote:
"A cycle between Rc pointers will never be deallocated. For this reason, Weak is used to break cycles. For example, a tree could have strong Rc pointers from parent nodes to children, and Weak pointers from children back to their parents."

So no. ```Rc``` in Rust as of today has no cycle detector. Maybe it did in the past when the language was younger. (Rust has undergone a lot of changes)

**Comment by clinuxrulz** [comment]
This is going to be challenging: https://doc.rust-lang.org/beta/std/rc/struct.Rc.html#method.get_mut
"Returns a mutable reference to the inner value, if there are no other Rc or Weak pointers to the same value."
We can only mutate data held by an ```Rc``` if no other ```Rc```s hold a reference to it.

**Comment by the-real-blackh** [comment]
In the Typescript version I implemented a cycle detector and that worked well. Inserting a weak reference in the right place is certainly a possible way of doing it. You just have to make sure it can never go wrong. I haven't thought through the implications.

**Comment by clinuxrulz** [comment]
I'll need to have more of a think as well. Its like the cell_loop/stream_loop node is split into two, lets call them the "Good Twin" and the "Evil Twin", the "Good Twin" always outlives the "Evil Twin", so its safe for a Weak Reference from the "Evil Twin" to the "Good Twin". And when we try to read the content of the "Evil Twin" we are always redirected to the "Good Twin" via the weak reference. I'm yet to think about what the implications are when ```switch``` is thrown into the mix.

**Comment by clinuxrulz** [comment]
I've started on a reference counted solution in the "refactor-rc" branch. It will take me quite a while to get it fully functional, so you may not hear from me for some time. I'll leave enough information in the data structures to allow for a cycle detector to be implemented if required. (I still think I can fool proof it with the technique in the last comment)

**Comment by the-real-blackh** [comment]
OK - good luck, and thanks for your work.
Note that the cycle detector I implemented in the typescript version is from a paper and is meant to be efficient. In the typescript version, the reference counting is done when you listen, because there's no way to count the number of times something is referenced in Javascript. The listen() operation was the only thing we could actually control. You won't need to do it this way.

**Comment by clinuxrulz** [comment]
OK I'm back... and wow, implementing a delayed stream for use in hold is a really mind bending, it was spinning me out for quite some time.

Thought I'd give you an update on where I was at.

The master branch is now the reference counted version, and 18 out of 19 tests are passing (all tests passing except switch_s)

```
test tests::constant_cell ... ok
test tests::apply ... ok
test tests::filter ... ok
test tests::hold ... ok
test tests::hold_is_delayed ... ok
test tests::lift ... ok
test tests::lift_from_simultaneous ... ok
test tests::loop_cell ... ok
test tests::lift_glitch ... ok
test tests::map ... ok
test tests::map_c ... ok
test tests::merge_non_simultaneous ... ok
test tests::merge2 ... ok
test tests::snapshot ... ok
test tests::merge_simultaneous ... ok
test tests::test_delay ... ok
test tests::values ... ok
test tests::switch_s ... FAILED
test tests::switch_c ... ok
```

I'll sort out ```switch_s``` later, I feel like just using the library for a while to try it out.

I'm currently a little bit confused about how long cells/streams should live for. Currently they live as long as the user holds a reference to the cell/stream, but I feel like it should be alive just a bit longer (maybe at least as long as the user has a reference to a listener of cell/stream).

If you imagine the graph being built from top to bottom, the strong reference point up the graph and the weak reference point down the graph at the moment, but may change.

How are lifetimes managed in the typescript version? I might be able to do the same.

One thought I was having is a "keep_alive_list" inside the cell_sinks/stream_sinks for any decendent nodes that have listeners, and then removing them from the keep_alive_list if their listeners are completely removed.
Pseudo Example:
```
{ <-- scope 1
  let s1 = new_cell_sink(1);
  { <-- scope 2
    let s2 = s1.map(|a| a + 1);
    s2.listen(|a| println!("{}", a));
  } <-- end of scope 2
  s1.send(1);
} <-- end of scope 1
```
In the above example a "keep_alive_list" can keep s2 alive for the lifetime of s1, so that "2" can be seen on the console output, even though s2 has gone out of scope.

The "keep_alive_list" may introduce a cycle, so it gets tricky here.

**Comment by clinuxrulz** [comment]
I also need to prune the graph periodically to remove any ```Weak``` references that no longer point to an object. I have not done this yet, but will do soon.

Rust has ```Rc``` for Strong Reference, and ```Weak``` for weak reference. When all strong references to a object are destroyed Rust will automatically mark the ```Weak``` reference pointing to the same object as invalid. I just need to search for invalid ```Weak``` references and remove them now and again.

**Comment by the-real-blackh** [comment]
In Java and similar implementations, cells and streams live while there is any reference to them, or if there is a listener registered on them.
In the Typescript implementation, cells/streams exist as Javascript objects but are not active, and can't process any input, until they're listened to via a call to listen(), which then constructs all the machinery. When anything is unlistened, it uses an algorithm from a paper to clean up cycles.
This approach is necessary because Javascript has no finalizers, so we just clean everything up when we find that the object is garbage collected like we do in Java and others.
I recommend you take a look at the Typescript version and see how it works, and consider translating it. The code is in Vertex.ts (because Node seems to clash with something in Typescript).

**Comment by clinuxrulz** [comment]
I had a pretty good look at your typescript version, and your code seems far easier to comprehend than what I have written. I'll start translating it entirely using a different branch and merge back into master if it works out.

**Comment by the-real-blackh** [comment]
You should have seen my first 10 attempts.
Once again, thank you so much for this work! I really appreciate it. I always thought Rust was a great language and this will be my excuse to learn it properly.

**Comment by clinuxrulz** [comment]
No problem. I'm really eager for a Sodium I can use for Rust.

**Comment by clinuxrulz** [comment]
I've been very gradually translating the typescript version in "ts-translation" branch.

source can be found here: https://github.com/SodiumFRP/sodium-rust/tree/ts-translation/src/sodium

I've translated Vertex.ts as Node.rs, and Translation.ts so far. And when I went to translate "Stream.ts", I've noticed these "Lambda1", "Lambda2", etc. that capture dependent Cells/Streams of the lambda. So I have a quick question:

Do I require these Lambdas that capture depend nodes in the Rust version? Or its it just a requirement for the TypeScript version?

In Rust we can increment/decrement the cycle detected reference counter (Node.rs) in the constructor/deconstructure of their owner rather than on listen/unlisten. And I'm not sure if this will make a difference concerning "Lambda1", "Lambda2", etc.

I'll keep translating as-is for now. We can still tweak how it works later on.

**Comment by the-real-blackh** [comment]
Hi,

I think Lambda1, etc, will be needed, because, for the cycle detection to work, we have to know what the dependencies are. Yes, it makes sense that you would do it in the constructor instead of listen. But I don't think that makes any difference. You still have to know what's inside the closure, otherwise you can't track the dependencies.

Steve


Steve

**Comment by clinuxrulz** [comment]
Thanks for your quick reply, I'll translate the Lambda1, 2, etc. as well.

**Comment by clinuxrulz** [comment]
Hi @the-real-blackh ,
I've made a mess of the typescript translation. Idiomatic typescript and idiomatic rust are different beasts. I pretty much used too many raw pointers confused myself. I also wanted to have another go at trying to do without a cycle detector, because I believe it can be done easily.

In the master branch is the start of the translation of the Java version of sodium. I believe I have a solution for the memory management. The trick is as follows:

```
pub struct Stream<A> {
    pub data: Rc<RefCell<StreamData<A>>>
}

pub struct WeakStream<A> {
    pub data: Weak<RefCell<StreamData<A>>>
}
```

A weak stream and a strong stream. A strong stream can be downgraded to a weak stream, and a weak stream can be upgraded to a strong stream. (corresponds to weak/strong reference).

Now in ```StreamLoop``` we simply make sure the closure that pushes value back only captures a ```WeakStream``` to push values too. All the exposed API can be made the same as the Java version. This should work even in the presents of a ```switchS```/```switchC```, because those methods really return a single ```StreamWithSend``` (out) which the switch to stream/cell tunnels its values through. ```switch``` should cause no issues with memory management. Also ```drop``` our deconstructor is like Java's finalizer.

The end user will not use ```WeakStream``` its just for internal implementation use. 

Clinton.

**Comment by the-real-blackh** [comment]
Hi @clinuxrulz,

I know this is difficult stuff! I am really grateful that you're doing this.

Just because StreamLoop is being used, it doesn't mean that a loop exists, because the value passed to myLoop.loop() may not depend on myLoop at all. StreamLoop and CellLoop are really only forward references. They don't guarantee loops.

The model could break if the dependency were only sometimes there because of a switch in the middle of the loop.

So, I think your method would work as long as you stipulate that the value passed to myLoop.loop() must depend on myLoop.

This might not be a terrible thing. The dependency tracking method used in Typescript requires all this declaring what's inside a closure, and it also places some restrictions on the use of .sample(). These are costs too, just like requiring the user to check the dependencies of a looped value. In Javascript the method you proposed wasn't an option because weak references don't exist.


Steve

**Comment by clinuxrulz** [comment]
+-----------------+
                                             |                 | (Always return the stream and value returned in the lambda passed to loop_s
                                             |                 V
     loop_s :: forall s a b. (Stream a -> (Stream a, b)) -> (Stream a, b)
                                ^                                |
                                |                                | (returned stream listen with weak reference to forward reference, and pushes values into it.)
                                +--------------------------------+

I wonder if a change in type signature as above can solve the problem. It should handle the case where a loop does not really exist.

**Comment by clinuxrulz** [comment]
@the-real-blackh, time for a little Magic!



Watch this:

```
public void loop(Stream<A> ea_out) {
    this.loop_(ea_out);
    return this.orElse(ea_out);
}
```

You can force myLoop.loop() to depend on myLoop regardless of what the end-user does, to make the memory management work properly.

**Comment by the-real-blackh** [comment]
It took me a little while. :) On languages where lambdas are not annoying to use, that approach is better than the one I use in Java and Typescript.

It still doesn't guarantee that the 'Stream a' you return depends on the 'Stream a' you were passed, though.

I know this is really pseudocode but if this is Haskell, you can't actually implement it with the exact type signature. There needs to be a monad of some kind, i.e. `(Stream a -> M (Stream a, b)) -> M (Stream a, b)`. In Reactive-Banana this monad is called Moment. The monad is needed by the hold primitive.

**Comment by clinuxrulz** [comment]
Reactive-Banana uses a ```MonadFix``` instance for its Moment Monad with the type signature.
```mfix :: (a -> m a) -> m a```.
So yeah, its really just a hack.

**Comment by clinuxrulz** [comment]
Hi Steve, a little update.

The Java version is completely translated. Value-wise all tests are passing, 3 tests have memory leaks. The memory leak ones are those involving ```CellLoop``` / ```StreamLoop```. I'm just going to go the ```.loop()``` and ```.ref()``` route and be done with it, it feels like a better option than having the end-user state their lambda's dependencies on sodium objects.
Slight module shuffle to do, so I can mimic package-private. Then I'll be done.

Clinton.

**Comment by clinuxrulz** [comment]
Something weird is going on, I've got ```weak_loop``` and ```strong_loop``` implemented for ```StreamLoop```.. But for ```CellLoop``` it seems that it just works with ```loop```, because it makes use of ```StreamLoop::weak_loop``` internally for the implementation of ```CellLoop::loop```.

So what if ```StreamLoop::loop``` can be implemented via ```CellLoop::loop``` internally? (keeping ```StreamLoop::weak_loop/strong_loop``` package private) Then it seems like all memory would be manage correctly without any special treatment from the end-user.

I'm not 100% certain of my claim. It will need some more investigation.

**Comment by clinuxrulz** [comment]
GitHub needs multi-page support for comments on issues. Its quite a scroll down to here.

I have a cleaner memory management hack for ```StreamLoop::loop```, using the ```or_else``` trick caused problems with ranks. Here is the new trick:

```
class StreamLoop<A> {
 . . .
    public void loop(Stream<A> ea_out) {
        this.loop_(ea_out);
        Stream<A> self = this;
        ea_out.unsafe_add_cleanup(new Listener() {
            private Stream<A> keepAlive = self;
            public void unlisten() {}
        });
    }
 . . .
}
```

And of course ```loop_``` makes a Weak link from ```ea_out``` back to ```this```.

**Comment by clinuxrulz** [comment]
Memory Management Problem... not solved :worried: . I've just realized you can make a loop without ```StreamLoop``` / ```CellLoop```. You can make a loop out of ```StreamSink``` with ```SwitchS```, and then the memory leak is back again.

Seems that TypeScript might have the better solution.

Current plan of attach, provide:

    weakS :: Stream a -> Stream a
    weakC :: Cell a -> Cell a

These ones behave as follows:

    let s2 = s1.weak(); // s2 now does not hold a strong reference to s1

Then let the end-user patch their memory leaks with ```weak```. Then gradually support the TypeScript solution as well, and leave it up to the end-user to decide on their approach.

I'm might be able to use https://github.com/Manishearth/rust-gc to speed up the process.

**Comment by the-real-blackh** [comment]
Sorry this is so difficult! If you can use rust-gc, I suggest you do. You will need finalizers. A gc is best overall from the point that it's absolutely impossible for the user of the library to get things wrong. With the typescript solution, the user has to declare all sodium objects held in closures. Also, sample() won't work correctly on cells that don't have listeners attached, which is a bit of a problem.

**Comment by clinuxrulz** [comment]
With rust-gc the Trace trait is required to be implemented for GC objects. So the end user will still need to declare sodium objects held in closures.

**Comment by clinuxrulz** [comment]
Change of plans. rust-gc does not support Weak references. So I translated the garbage collector from the paper "A Pure Reference Counting Garbage Collector". (Same one used by TypeScript)

https://github.com/SodiumFRP/sodium-rust/blob/master/src/sodium/gc.rs

Hopefully I can just replace all the ```Rc<>```s with ```Gc<>``` and be done with it.

**Comment by the-real-blackh** [comment]
I think we're probably stuck with this approach in Rust, C++ and Typescript. It may require a little bit of experience to find corner cases like this one: https://github.com/SodiumFRP/sodium-typescript/issues/42
But, I think this way of doing things is both robust and efficient, and the challenges are solvable.

**Comment by clinuxrulz** [comment]
Closing issue here, because its getting to long.

I'm finished with it, and I'll let ya know somewhere else when its done.

Two things I've noticed:
1) If the end user forgets to mark a dependency, then there is a chance of a leak.
2) If the end user marks something as a dependency that is not a dependency, then there is a chance of a segfault.

TypeScript avoids the chance of segfault. GC on top of GC.

I think here we will just put faith in the end user. Things are complicated enough already.
